docs/content/3.middleware/2.guides/10.working-with-logger.md

```diff
@@ -0,0 +1,336 @@
+
+## Using logger
+
+The middleware application provides access to the logger in various parts of the system, such as in extensions and integrations. This flexibility allows you to log important events, errors, and other data throughout the lifecycle of your application. Whether you’re working in an extension or an integration, you can easily utilize the logger to capture and track necessary information.
+
+```json[Example log]
+{
+  "message": "log message we print",
+  "timestamp": "2024-10-22T19:04:18.791Z",
+  "severity": "INFO",
+  "context": "middleware",
+  "scope": {
+    "integrationName": "adyen",
+    "functionName": "addedCustomEndpoint",
+    "hookName": "afterCall",
+    "type": "requestHook",
+    "extensionName": "testing-adyen-extension"
+  }
+}
+```
+
+In this section, we will explore how to access and use the logger in different parts of the application, ensuring you have the tools to log data efficiently and consistently. 
+
+These examples will demonstrate the proper ways to obtain and interact with the logger, ensuring that you are able to implement consistent and effective logging practices across your middleware application. As you will see, `getLogger` function solves the problem in most cases.
+
+### Example log
+
+### extendApp
+
+```ts
+import { getLogger } from "@vue-storefront/middleware";
+
+const someExtension = {
+  name: "some-extension",
+  extendApp(params) {
+    const logger = getLogger(params);
+    logger.info("Logged succesfully!");
+  }
+};
+```
+
+### hooks
+
+Logger is available in hooks factory:
+
+```ts
+import { getLogger } from "@vue-storefront/middleware";
+
+const someExtension = {
+  name: "some-extension",
+  hooks(req, res, alokai) {
+    const logger = getLogger(alokai);
+    logger.info("hooks");
+    return {
+      // ...
+    };
+  },
+};
+```
+
+In order, to use logger in certain hooks like `beforeCall`, `beforeCreate`, `afterCall`, `afterCreate` obtain it from it's params:
+
+```ts
+import { getLogger } from "@vue-storefront/middleware";
+
+const someExtension = {
+  name: "some-extension",
+  hooks(req, res, alokai) {
+    return {
+      beforeCall(params) {
+        const logger = getLogger(params);
+        logger.info("smth");
+        return params.args;
+      },
+      beforeCreate(params) {
+        const logger = getLogger(params);
+        logger.info("smth");
+        return params;
+      },
+      afterCall(params) {
+        const logger = getLogger(params);
+        logger.info("smth");
+        return params.response;
+      },
+      afterCreate(params) {
+        const logger = getLogger(params);
+        logger.info("smth");
+        return params;
+      },
+    };
+  },
+};
+```
+
+#### Caveat: Accessing logger via closure
```

Comment:
I did discuss it with Bartek, I honestly don't know why do we have to have this as a function and the argument that "well we may want to put it in a different place later" doesn't really speak to me. It should either be context or req.
Think about this that way - now user has to import it to even use it, rather than simply access is, so it increases the chances that the person will use `console` instead of our logger.

`this` would be a new convention in our product, so I'm against this.


I won't die on this hill tho, the logger was in a development already for quite some time.

Reply:
Agree

Reply:
Firstly, do not use colorized sentences like "well we may want to put it in a different place later". You put us in a place where we do not have arguments, where it is not true. It is neither a quote, neither our position in this case. We have provided a logic arguments and you respond with what you think which is a word against word at best.

1. it is not possible to logger only on the context or req (req is unusable at all and we must use res as only res have locals to store some custom data).
2. in some places only context is available, in some only res
3. in some places, like hooks, which is a great example for you, we can't use neither of them because of race conditions in hooks execution. We can't inject logger into singletons.
4. we have discussion with DX about the approach and we showed alternatives
5. logger is set on:

- res.locals.alokai.logger
- context.logger
- someObject.res
- container.logger (injected into hooks to avoid race conditions issues)

So the user will have to remember all these use cases to be able to get the logger.
Also, if we change the path, it will be a breaking change for all users and for all our packages. 

To address these issues we introduced getLogger function that hide all that complexity which which, in addition, prevents from bc.
Even now your request implies changes in 20 packages at least, and only 20 because we did not add logger everywhere yet.

Reply:
I wasn't attempting a strawman, sorry if it looked this way. This was the final argument I remembered from our call after the conversation we've had. Not exactly in this phrasing but the summary of it.
- The fact that req is unusable I don't remember at all, if you said it it wasn't back on our calls. 
- Putting it on res is something completely new to me.
- We already had a discussion that there don't have to be any singletons, we have a full control over what do we pass to parameters of hooks and methods, we can build objects on the fly.

That being said, it doesn't really matter at the end of the day. We would go back and argue in circles again and it is greenlighted already.

Reply:
`The fact that req is unusable I don't remember at all, if you said it it wasn't back on our calls.` 
You can't remember that because we did not know that at that time.
`We already had a discussion that there don't have to be any singletons, we have a full control over what do we pass to parameters of hooks and methods, we can build objects on the fly.`
But the req/res object is and will be a singleton - and must be a singleton. And the same req/res is used in parallel. We have control of what we pass but we can't pass it into these object in all places (eg. in hooks, we just cant due to condition race).

We have considered `this.logger` but it is far from being a perfect solution. It will not work for arrow functions which is the major of our code, we can't bind this. Also, using `this` in such way is an antipattern.

We can also consider attaching getLogger to the global object - which is also an anti pattern but, at least, it is not "this".

Reply:
@filrak Indeed, it would be pleasant to have it always available via `this`. However, we cannot bind `this` for arrow functions. Example (last 2 lines are most important):
```ts
const obj = {
  value: 'Object',
  regularFunction: function() {
    console.log(this.value);
  },
  arrowFunction: () => {
    console.log(this.value);
  }
};

obj.regularFunction(); // Will print 'Object'
obj.arrowFunction(); // Will print 'undefined' in the browser or error in Node.js

// New binding of `this` with `bind`
const newRegularFunction = obj.regularFunction.bind({ value: 'New Object' });
newRegularFunction(); // Will print 'New Object'

const newArrowFunction = obj.arrowFunction.bind({ value: 'New Object' });
newArrowFunction(); // Will still print 'undefined' in the browser or error in Node.js

```

So this change would require rewriting every handler using logger in every integration/module/extension to be classic non-arrow function. We would also need to have constraint to only use classic functions in some places of our software. 

What's more, it doesn't solve the problem of accessing the wrong logger via closure, e.g:
```ts
const someExtension = {
  name: "some-extension",
  hooks(req, res, alokai) {
    // const hooksLogger = getLogger(alokai);
    const hooksLogger = this.logger;
    hooksLogger.info("hooks");
    return {
      beforeCall(params) {
        // Never access via closure a logger belonging to the hooks factory:
        hooksLogger.info("smth");
        // Instead use own logger of every hook function:
        const logger = this.logger;
        // const logger = getLogger(params);
        logger.info("smth");
        return params.args;
      }
    };
  },
};
```

Which unfortunately, invalidates an idea of passing logger via `this`.

BTW I don't understand mentioning LoggerManager. It's not public thing anymore, it's used only internally. So end-developer won't be overwhelmed by new thing to learn.

@michaelKurowski Yup, we learned a lot more during work on the logger and haven't had a chance yet to tell about the details of our findings. 

> I wasn’t aware that there are places exposed to the user that only have res available, looking at the code now, looks like you probably refer to the after hook?

Bartek used mental shortcut, he meant we have access to either context or req+res pair.

BTW At first, I also thought `after hook` has access to the `response: Response` from express. However, it has value of response.body. 

### Our final approach 
It mostly involves injecting a logger wrapped with proper metadata in every possible place. Every place has its own set of metadata attached to make it resistant to singleton-related issues and race conditions. We couldn't set it in `context.res` and attach metadata on the fly, as it causes issues in case of [parallel calls of multiple handlers in the scope of one request](https://docs.alokai.com/middleware/guides/orchestration#example-1-fetching-custom-product-properties-from-legacy-systems) and would make code really hard to grasp.

I didn't want to be too overwhelming with explanation so I've cut to the chase. If you wish to know more about challenges, tell me, as I am open to writing additional longer comment.

Reply:
Can you elaborate more on this final approach? 

I get the argument to not use `this` and indeed one less import is not worth refactoring all packages. Though If we can still have it available directly from the function parameters instead of relying on `getLogger` import and having possibility to use it wrong when we don't inject it on every level would be a step forward in reducing complexity. Can we have that?


Also" it's used only internally. So end-developer won't be overwhelmed by new thing to learn.". Wrong. Integration API is also a public API. Customers often write their own integrations with various systems.

Reply:
I will let @Fifciu to answer it but:
`Though If we can still have it available directly from the function parameters instead of relying on getLogger`
the logger itself is in very different places depending on if this is a hook, extension, or other extensibility place. End user will have to remember where and how to get it for all these places. `getLogger` hides this complexity.
`and having possibility to use it wrong when we don't inject it on every level would be a step forward in reducing complexity.` I do not understand the rest of it; What possibility to use it wrong? Where we do not inject the logger? Could you provide some examples where `getLogger` leads to issues and function-less approach does not?

As I described, here are 3 places where logger can be accessed, depends on the extension point:

res.locals.alokai.logger - if res if available and we do not have to care about race condition
context.logger - if context is available but res is not
alokai.logger - injected into hooks to avoid race conditions issues

`getLogger` not only hides that complexity, protect clients from bc but also help with types inference - it would be a hard or impossible to rework broken typing in middleware in such way that the logger will be properly typed in all these places.

`Also" it's used only internally. So end-developer won't be overwhelmed by new thing to learn.". Wrong. Integration API is also a public API. Customers often write their own integrations with various systems.` This is about something completely different, Filip was referring to some internal things which are irrelevant here. We still use loggerManager internally.

Reply:
@filrak This is our final list of arguments, if you do not think that these are valid points we will do as you ask.

- Users do not need to remember where to obtain the logger, as it is encapsulated within the function itself.
- The function can provide documentation within a docblock, immediately clarifying the objects it operates on (beyond TypeScript types). This eliminates the need to leave the IDE to refer to external documentation.
- By encapsulating specific logic within the function, we make it easier to modify in the future and protect clients from breaking changes.
- This approach enables us to maintain proper typing, which might not be achievable in our middleware under other circumstances (to be validated).
- Every method of accessing the logger must be documented. When using a function, the docblock and brief in-documentation note provide clear information, ensuring comprehensive documentation.

Reply:
@filrak @michaelKurowski 
Guys, what is your final decision? Can we use `getLogger` or we should get rid of it from the public interface?

Reply:
Thanks for giving the arguments, they are convincing.

Frankly I think given your arguments choosing one or the other will result in tradeoffs coming purely from the weakness of our implementation rather than thinking what is the best possible API for this kind of feature, and I'm sure there is an approach that won't have them (some form of context-aware `getLogger` from fn context/params ). Though, if you don't see such solution at first glance lets proceed with your proposal.